django/middleware/csrf.py

```diff
@@ -227,15 +257,21 @@ def process_view(self, request, callback, callback_args, callback_kwargs):
                     return self._reject(request, REASON_INSECURE_REFERER)
 
                 # If there isn't a CSRF_COOKIE_DOMAIN, assume we need an exact
-                # match on host:port. If not, obey the cookie rules.
-                if settings.CSRF_COOKIE_DOMAIN is None:
-                    # request.get_host() includes the port.
-                    good_referer = request.get_host()
-                else:
+                # match on host:port. If not, obey the cookie rules (or those
+                # for the session cookie, if we use sessions).
+                if settings.CSRF_COOKIE_DOMAIN is not None and not settings.CSRF_USE_SESSIONS:
                     good_referer = settings.CSRF_COOKIE_DOMAIN
                     server_port = request.get_port()
                     if server_port not in ('443', '80'):
                         good_referer = '%s:%s' % (good_referer, server_port)
+                elif settings.SESSION_COOKIE_DOMAIN is not None and settings.CSRF_USE_SESSIONS:
+                    good_referer = settings.SESSION_COOKIE_DOMAIN
+                    server_port = request.META['SERVER_PORT']
```

Comment:
Is there any reason you are not using `request.get_port` like above? I also do not like the duplication, can you set `good_referer` before like:
```
good_referer = settings.SESSION_COOKIE_DOMAIN if settings.USE_CSRF_SESSIONS else settings.CSRF_COOKIE_DOMAIN
```
and then simply check on
```
if good_referer is not None and …
```